### ISSUE TITLE
[DevicePlugin] Change the multiple Allocate call to a single Pod Admit call

### ISSUE THREAD

**Issue by RenaudWasTaken** [kubernetes/kubernetes#59370]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature
/sig node
/area hw-accelerators
/cc @vishh @jiayingz @vikaschoudhary16 

**What happened**:
At the moment device plugin Allocate call is called once for each container at Pod admition time.
As we introduce the Initialize call, this behavior is no longer needed.
This behavior also leads to weird API semantics and makes it really hard for the device plugin to track what device is Allocated and what device isn't (related: [Deallocate](https://github.com/kubernetes/kubernetes/issues/59110), [Partial allocation failures](https://github.com/kubernetes/kubernetes/issues/56307)).

**What you expected to happen**:
We could change the lifecycle of this call to a single call at Pod admition time that would gather all the needed runtime setup requirements.

It is also a good foundation for integrating the [NUMA manager](https://github.com/kubernetes/community/pull/1680) with the device plugin API as well as passing [additional information to the device plugin](https://github.com/kubernetes/kubernetes/issues/59109) depending on what agreement we arrive at.

In general it makes the lifecycle model cleaner as we have one call at pod admission (`PodAdmit`?) and one call before each container start (`InitializeContainer`?).

It will also pave the way for a better sharing API as the device plugin will have a better view of what is being shared at the pod level (eg: NICs) rather than try to guess what device gets shared at the container level (though this might not apply for all devices, e.g: GPUs).

The gRPC API might look like the following:
```protobuf
// AdmitPodRequest is a call issued by the kubelet during pod admission
// TODO this message should be used by the CPU manager
// TODO the first two fields might be a good answer for  issue #59109
message AdmitPodRequest {
        // Name of the pod
        string podname = 1;

        // Annotations on the pod
        map<string, string> annotations = 2;

        // list of InitContainers with the assigned devices
        repeated Container init_containers = 3;

        // list of Containers with the assigned devices
        repeated Container containers = 4;
}

message AdmitPodResponse {
        // map<containerName, containerRuntimeSpec>
        map<string, RuntimeSpec> specs = 1;
}

message Container {
        // The name of the Container
        string name = 1;
        // The device assigned to that container
        repeated string devices = 2;
}

message RuntimeSpec {
        // List of environment variable to be set in the container to access one of more devices.
	map<string, string> envs = 1;
	// Mounts for the container.
	repeated Mount mounts = 2;
	// Devices for the container.
	repeated DeviceSpec devices = 3;
}
```

**Comment by lichuqiang** [comment]
```
// AdmitPodRequest is a call issued by the kubelet during pod admission
// TODO this message should be used by the CPU manager
// TODO the first two fields might be a good answer for  issue #59109
message AdmitPodRequest {
        // Name of the pod
        string podname = 1;

        // Annotations on the pod
        map<string, string> annotations = 2;

        // list of InitContainers with the assigned devices
        repeated Container init_containers = 3;

        // list of Containers with the assigned devices
        repeated Container containers = 4;
}
```
That would be a big change, 
so you mean we should rely on the plugin itself to decide which devices to allocate? IMO a plugin should not be aware of the logic inside a pod when managing allocation(i.e. devices of a init container could be reused by regular containers and other containers)

**Comment by RenaudWasTaken** [comment]
> That would be a big change,
> so you mean we should rely on the plugin itself to decide which devices to allocate? 

No, I'm suggesting we merge the N allocation calls into one call at pod admission.

> IMO a plugin should not be aware of the logic inside a pod when managing allocation (i.e. devices of a init container could be reused by regular containers and other containers)

That's going to be a requirement for GPU sharing (e.g: [MPS needs a daemon to run](https://docs.nvidia.com/deploy/pdf/CUDA_Multi_Process_Service_Overview.pdf)) and in general pod level sharing of devices.
Any reasons or arguments to support your opinion that I might not have considered?

**Comment by lichuqiang** [comment]
>No, I'm suggesting we merge the N allocation calls into one call at pod admission. Kubelet would still be deciding which container gets which device.

Oh, I have no objection if you suggest things this way. It is because I didn't see input of device info in struct `AdmitPodRequest` above that I thought you were suggesting moving the decision part to plugins.

**Comment by vikaschoudhary16** [comment]
I am just thinking if passing container name could be a concern. Thinking more on it.

**Comment by jiayingz** [comment]
@RenaudWasTaken for MPS use cases, we would like to learn more on its explicit requirements for Kuberentes. Do you have any doc that describes these requirements and the high-level structure?

**Comment by RenaudWasTaken** [comment]
> @RenaudWasTaken for MPS use cases, we would like to learn more on its explicit requirements for Kuberentes. Do you have any doc that describes these requirements and the high-level structure?


Reposting the link in the previous post:  https://docs.nvidia.com/deploy/pdf/CUDA_Multi_Process_Service_Overview.pdf

**Comment by RenaudWasTaken** [comment]
> Agree with Vikas' concerns. I think other kubelet plugins, like CSI, have been explicit on avoid passing Pod and Container information to plugins. Device plugin probably want to follow the similar model. I think this also allows better separation of concerns that device plugins only need to care about devices.

API-wise I don't think it's a good model to rely on the table ordering.

In terms of general model, we've already seen multiple asks (from basically every device plugin author except GKE) for passing container and pod information, I don't see how we're going to escape passing none.
@vikaschoudhary16 even reads the checkpointing file from the device plugin to get the pod name....

Now in terms of Init Containers the behavior is different from normal containers, I don't see that as a good model to still merge them as a normal container list.
This is going to affect basic sharing (sharing a GPU between multiple containers) and more advanced sharing (i.e MPS).

**Comment by jiayingz** [comment]
I know there have been multiple asks on passing more container and pod information to plugins. I think other K8s areas have also seen similar asks but they have avoided doing this for good reasons. We are going to see more plugins on different areas like devices, storage, and network, and will need a general policy on what information should be passed to plugins. If you believe there are strong reasons to pass pod and container information to plugins, I suggest you to discuss this with sig-node leads and other senior K8s leads. Until we get consensus and green lights from them, we are not going to open the door on device plugins.

For Solarflare device plugins, I think using checkpoint file is meant to be a temporary solution to unblock their progress. I don't know the specific reasons why they don't want to use CNI to expose network interfaces. It could be CNI is hard to use. If that is the case, I think there have been discussions on a network plugin model that could help address this problem. When we were discussing a general kubelet plugin registration model, we also discussed how to make it easy to support multiple types of plugins from a single component. In terms how to make it easy to expose a network interface on K8s, I think sig-network perhaps a better place for such topics.

For device sharing, I believe it is very important but unfortunately we are quite far away from supporting it on K8s today. We definitely would like to help on coming up with a long-term plan to support this. To get started, we need to better understand device sharing requirements. The MPS overview doc you shared is very helpful for us to learn the general Nvidia GPU MPS model. What we would also like to see is a document on the specific GPU sharing use cases with containers, the requirements on K8s, and the general model how you expect things to work from Nvidia point of view. Until we learn such details, just saying this is needed by MPS will not help us evaluate whether we should implement a feature on K8s.

**Comment by RenaudWasTaken** [comment]
Ok let's move the discussion of device informations to #59109

I'll also create an issue for supporting device sharing which for basic sharing (i.e two containers using the same device) or pod level sharing (i.e: NICs) is really not far away.
It only requires consensus on the Device Plugin API and the User facing API.

Let's focus on the current suggested change:
- Do we support merging the current Allocate call into one single call
- Do we have a problem with the container name being passed down to the device plugin?
  - What it brings is a clear explicit API that doesn't depend on the un-explicit ordering.
- Do we have a problem with separation of init container and container separation?
  - I realize this gives "some" pod information but it seems sane to me from an API standpoint that a device plugin might not be performing the same actions on init containers than on normal containers.
